components/support/migration/src/main/java/mozilla/components/support/migration/GeckoMigration.kt

```diff
@@ -17,27 +94,73 @@ internal object GeckoMigration {
      * Tries to migrate Gecko related internal files in [profilePath].
      */
     fun migrate(
-        profilePath: String
-    ): Result<Unit> {
+        profilePath: String,
+        migrationVersion: Int
+    ): Result<GeckoMigrationResult> {
         // GeckoView will happily pick up the profile from Fennec and reuse all data in it. So this
-        // migration is mostly focused on removing the files that we do not want to reuse.
+        // migration is mostly focused on removing all prefs that we do not want to reuse.
+        val prefsjs = File(profilePath, PREFS_FILE)
+        if (!prefsjs.exists()) {
+            return Result.Success(GeckoMigrationResult.Success.NoPrefsFile)
+        }
+
+        // First, let's create a backup of the file.
+        try {
+            prefsjs.copyTo(File(profilePath, PREFS_BACKUP_FILE + migrationVersion), true)
+        } catch (e: IOException) {
+            return Result.Failure(GeckoMigrationException(GeckoMigrationResult.Failure.FailedToWriteBackup(e)))
+        }
 
-        // The prefs file contains gecko prefs that have been changed by the user (about:config) or
-        // code. Since those may be incompatible with GeckoView and/or app code, we decided to just
-        // reset them. GeckoView will recreate the file with defaults if needed.
-        return removePrefsFile(File(profilePath))
+        // Now let's transform the original to contain only prefs we want to keep.
+        val transformed = prefsjs.useLines {
+            it.filter { line ->
+                userPrefsToKeep.any { pref -> line.startsWith("$USER_PREF(\"$pref\"") }
+            }.toList()
+        }
+
+        return if (transformed.isEmpty()) {
+            removePrefsFile(prefsjs)
+        } else {
+            // Run a very basic validation and fall back to removing the file if we find anything wrong.
+            val invalid = transformed.any { line -> !isLineValid(line) }
+            if (invalid) {
+                removePrefsFile(prefsjs, invalid = true)
+            } else {
+                rewritePrefsFile(prefsjs, transformed)
+            }
+        }
     }
 
-    private fun removePrefsFile(profilePath: File): Result<Unit> {
-        val prefsFile = File(profilePath, PREFS_FILE)
-        if (prefsFile.exists()) {
-            return if (prefsFile.delete()) {
-                Result.Success(Unit)
+    private fun isLineValid(pref: String): Boolean {
+        return pref.lastIndexOf("$USER_PREF") == 0 && pref.endsWith(");")
+    }
+
+    @Suppress("TooGenericExceptionCaught")
+    private fun removePrefsFile(prefsJs: File, invalid: Boolean = false): Result<GeckoMigrationResult> {
+        return try {
+            if (prefsJs.delete()) {
```

Comment:
if you care about logging the exact exception: ` Note that the Files class defines the delete method to throw an IOException when a file cannot be deleted. This is useful for error reporting and to diagnose why a file cannot be deleted.`

Reply:
Yeah, we're already catching the exception, but in case it returns false without throwing we also report that as an error. Or did I misunderstand what you meant?

Reply:
Oh, I see you suggest using `Files.delete` instead.